gadgets/src/permutation.rs

```diff
@@ -0,0 +1,312 @@
+            // skip first fingerprint for its prev_fingerprint
+            if offset != 0 {
+                self.q_row_non_first.enable(region, offset)?;
+            }
+
+            self.q_row_enable.enable(region, offset)?;
+
+            let row_acc_fingerprint_cell = region.assign_advice(
+                || format!("acc_fingerprints at index {}", offset),
+                self.acc_fingerprints,
+                offset,
+                || *row_acc_fingerprints,
+            )?;
+
+            region.assign_advice(
+                || format!("row_fingerprints at index {}", offset),
+                self.row_fingerprints,
+                offset,
+                || *row_fingerprints,
+            )?;
+
+            let alpha_cell = region.assign_advice(
+                || format!("alpha at index {}", offset),
+                self.alpha,
+                offset,
+                || alpha,
+            )?;
+            let gamma_cells = self
+                .power_of_gamma
+                .iter()
+                .zip(power_of_gamma.iter())
+                .map(|(col, value)| {
+                    region.assign_advice(
+                        || format!("gamma at index {}", offset),
+                        *col,
+                        offset,
+                        || *value,
+                    )
+                })
+                .collect::<Result<Vec<AssignedCell<F, F>>, Error>>()?;
+
+            if offset == 0 {
+                alpha_first_cell = Some(alpha_cell);
+                gamma_first_cell = Some(gamma_cells[0].clone());
+                acc_fingerprints_prev_cell = Some(row_acc_fingerprint_cell.clone());
+            }
+            // last offset
+            if offset == fingerprints.len() - 1 {
+                last_fingerprint_cell = Some(row_acc_fingerprint_cell);
+            }
+        }
+
+        Ok((
+            alpha_first_cell.unwrap(),
+            gamma_first_cell.unwrap(),
+            acc_fingerprints_prev_cell.unwrap(),
+            last_fingerprint_cell.unwrap(),
+        ))
+    }
+
+    /// Annotates columns of this gadget embedded within a circuit region.
+    pub fn annotate_columns_in_region(&self, region: &mut Region<F>, prefix: &str) {
+        [
+            (
+                self.acc_fingerprints,
+                "GADGETS_PermutationChipConfig_acc_fingerprints".to_string(),
+            ),
+            (
+                self.row_fingerprints,
+                "GADGETS_PermutationChipConfig_row_fingerprints".to_string(),
+            ),
+            (
+                self.alpha,
+                "GADGETS_PermutationChipConfig_alpha".to_string(),
+            ),
+        ]
+        .iter()
+        .cloned()
+        .chain(
+            self.power_of_gamma
+                .iter()
+                .enumerate()
+                .map(|(i, col)| (*col, format!("GADGETS_PermutationChipConfig_gamma_{}", i))),
```

Comment:
I find it clearer if we start counting on `1` so that the `ith` column represents `gamma^i`.